src/halmos/mapper.py

```diff
@@ -83,21 +84,123 @@ class BuildOut(metaclass=SingletonMeta):
+            for link in links:
+                start = link["start"]
+                end = start + link["length"]
+                placeholders.append((start, end))
+
+        for libs in deployed.get("linkReferences", {}).values():
+            for links in libs.values():
+                for link in links:
+                    start = link["start"]
+                    end = start + link["length"]
+                    placeholders.append((start, end))
+
+        placeholders = sorted(placeholders, key=lambda x: x[0])
+
+        # sanity check
+        last = 0
+        for start, end in placeholders:
+            if not (last <= start and start < end):
+                raise ValueError("invalid placeholders")
+            last = end
+        if last > len(deployed["object"][2:]) // 2:
+            raise ValueError("invalid placeholders")
+
+        return placeholders
+
+    def create_build_out_map_code(self):
+        """
+        Creates a mapping between deployed bytecode and contract names.
+
+        This mapping utilizes deployed bytecode because it retains its original length,
+        unlike creation bytecode which can be expanded by constructor arguments.
+
+        Since compile-time bytecode may contain placeholders for immutables or library addresses,
+        the actual deployed bytecode can differ from its compile-time code.
+        To accommodate this, the mapping is constructed from bytecode size to a list of tuples containing bytecode and contract name.
+        When querying this mapping, it first retrieves a list of tuples based on their size,
+        and then iterates through the list, comparing each bytecode.
+        """
+        self._build_out_map_code = defaultdict(list)
+
+        for filename, file_map in self._build_out_map.items():
+            for contract_name, (contract_map, _, _) in file_map.items():
+                deployed = contract_map["deployedBytecode"]
+                hexcode = deployed["object"][2:]  # drop '0x' prefix
+                if not hexcode:
+                    continue
+
+                size = len(hexcode) // 2  # byte length
+                placeholders = self.get_placeholders(deployed)
+                code_data = (hexcode, placeholders, contract_name, filename)
+                self._build_out_map_code[size].append(code_data)
+
+    def get_by_code(self, bytecode: ForwardRef("ByteVec")) -> tuple:
+        """
+        Return the contract name and file name of the given deployed bytecode.
+        """
+        if not self._build_out_map_code:
+            self.create_build_out_map_code()
+
+        # compares the deployed bytecode with the compile-time hexcode, excluding placeholders from the comparison
+        def eq_except_placeholders(hexcode: str, placeholders):
+            last = 0
+            for start, end in placeholders:
+                if not eq_bytes(bytecode[last:start], hexcode[2 * last : 2 * start]):
```

Comment:
I see, we have to do the indexing dance because for immutables placeholders look like `0000...0000` but for libraries they look like `__$72fd9f18565b1bf49af679aa1eb458ccdd$__`

Reply:
yes exactly.